AI/src/search/mcts/selection/PNS_UCB1.java

```diff
@@ -0,0 +1,429 @@
+        	{
+                if (child.disproofNumber() == 0 && child.numVisits() > minVisitsSolvedChild) 
+                	continue;
+            }
+        	
+        	final double exploit;
+        	final double explore;
+
+        	if (child == null)
+        	{
+        		exploit = unvisitedValueEstimate;
+        		explore = Math.sqrt(parentLog);
+        	}
+        	else
+        	{
+        		exploit = child.exploitationScore(moverAgent);
+        		final int numVisits = Math.max(child.numVisits() + child.numVirtualVisits(), 1);
+        		explore = Math.sqrt(parentLog / numVisits);
+        	}
+        	
+        	final double pnsTerm = currentPNMCTSNode.childrenPNSSelectionTerms()[i];
+
+        	final double ucb1Value = exploit + explorationConstant * explore + pnConstant * pnsTerm;
+        	//System.out.println("ucb1Value = " + ucb1Value);
+        	//System.out.println("exploit = " + exploit);
+        	//System.out.println("explore = " + explore);
+
+        	if (ucb1Value > bestValue)
+        	{
+        		bestValue = ucb1Value;
+        		bestIdx = i;
+        		numBestFound = 1;
+        	}
+        	else if 
+        	(
+        		ucb1Value == bestValue 
+        		&& 
+        		ThreadLocalRandom.current().nextInt() % ++numBestFound == 0
+        	)
+        	{
+        		bestIdx = i;
+        	}
+        }
+        
+        return bestIdx;
+	}
+	
+	//-------------------------------------------------------------------------
+	
+	/**
+	 * Updates the PN-based terms of the selection strategy for all children of current.
+	 * @param current
+	 */
+	public void updateChildrenSelectionScores(final PNMCTSNode current)
+	{
+		//  This is the array that we'll re-compute
+		final double[] childrenPNSSelectionTerms = current.childrenPNSSelectionTerms();
+		final PNMCTSNodeTypes currentNodeType = current.nodeType();
+		final int numLegalMoves = current.numLegalMoves();
+		
+		switch(pnsVariant)
+        {
+            case RANK:
+            	final List<ScoredInt> sortedChildIndices = new ArrayList<ScoredInt>();
+            	
+            	if (currentNodeType == PNMCTSNodeTypes.AND_NODE)
+            	{
+            		// AND node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: disproof number = 1.0 for unexpanded child. TODO this correct?
+							sortedChildIndices.add(new ScoredInt(i, 1.0));
+						}
+						else
+						{
+							sortedChildIndices.add(new ScoredInt(i, child.disproofNumber()));
+						}
+					}
+            	}
+            	else
+            	{
+            		// OR node
+            		for (int i = 0; i < numLegalMoves; ++i)
+					{
+						final PNMCTSNode child = (PNMCTSNode) current.childForNthLegalMove(i);
+						if (child == null)
+						{
+							// This means: proof number = 1.0 for unexpanded child. TODO this correct?
```

Comment:
Jakub please check if this is what we want for children that have not been added to tree yet